### ISSUE TITLE
Clean make dependencies after build

### ISSUE THREAD

**Issue by pauldotknopf** [trizen/trizen#107]
If I install packages from rust or go that need to be compiled, I'd like to remove rust/go when complete.

**Comment by trizen** [comment]
For this we have:
```console
trizen -Rcns $(pacman -Qqdt)
```

**Comment by fancytenseletters** [comment]
@trizen are you sure users want to execute command with that many options needed? It's worthless as it's actually easier to use Pacman alone.

Meanwhile yay has `yay -c`  option which does everything above. I think Trizen users will benefit from something comparable.

**Comment by stef204** [comment]
@trizen Not sure this can be reopened but I clearly second that feature (removing make dependencies after installing).
It is the only feature I find missing from Trizen.  And would be a valuable addition.
(This may be a matter of opinion, but just dumping make deps in one's system continuously, then cleaning them up with `trizen -Rcns $(pacman -Qqdt)` and` hoping it catches everything does not seem good practice.)

**Comment by stef204** [comment]
@trizen Not sure this can be reopened but I clearly second that feature (removing make dependencies after installing).
It is the only feature I find missing from Trizen.  And would be a valuable addition.
(This may be a matter of opinion, but just dumping make deps in one's system continuously, then cleaning them up with `trizen -Rcns $(pacman -Qqdt)` and` hoping it catches everything does not seem good practice.)

**Comment by trizen** [comment]
I would hardly call this a "feature". Automatically removing stuff from the system is a bad idea on its own.

Regarding `trizen -Rcns $(pacman -Qqdt)`, there is no hope involved.

Alternatively, there is `trizen --stats`, which prints the unneeded packages on the last line of the output, which can be copy-pasted after `trizen -Rcns `.

**Comment by pauldotknopf** [comment]
I'd say build your packages in a container.

**Comment by Kr1ss-XD** [comment]
> It is the only feature I find missing from Trizen. And would be a valuable addition.

Why not just put 
```
makepkg_command  =>  "/usr/bin/makepkg -src"
```
your `trizen.conf` ?  The `-r`/`--rmdeps` flag should exactly do what you are missing.

**Comment by stef204** [comment]
> ```
> makepkg_command  =>  "/usr/bin/makepkg -src"
> ```
> 
> in your `trizen.conf` ? The `-r`/`--rmdeps` flag should exactly do what you are missing.

Great thinking!  I am just migrating to trizen so still trying to get familiar with it. Didn't think/notice I could edit the pacman command inside trizen.conf.

Thanks!

**Comment by Morganamilo** [comment]
And what if your makedeps have deps?

**Comment by Kr1ss-XD** [comment]
Deps of deps are still deps, so they would be removed after the builld if they weren't installed before, otherwise makepkg wouldn't touch them.

**Comment by Morganamilo** [comment]
Yes but makepkg isn't the one installing aur makedeps.

**Comment by Kr1ss-XD** [comment]
Ah okay, I see. So you would still have to take care of those manually. :disappointed:

**Comment by stef204** [comment]
> I would hardly call this a "feature". Automatically removing stuff from the system is a bad idea on its own.

`makepkg -sr ` or `makepkg -sri` from the directory containing the PKGBUILD does the job.
If a package needs a make dep which is not installed, it will install it, build and remove it after the build.  Perfectly acceptable.

@Kr1ss-XD provided the input needed to use it with trizen.

**Comment by stef204** [comment]
I may have spoken to soon.

I modified the default `makepkg` command to `makepkg -srcf` but installing adobe-air caused trizen (or makepkg) to install the deps for adobe-air-sdk, then it removed them, then reinstalled them, installed the last package then removed them again.

So either the `-srcf` options are not the right ones (I kept the `-c` in there so that it cleans up after itself, and the `-f` as well to try to stay close to the trizen default.)

Something's amiss.

It does seem that for the sake of simplicity, @trizen's  approach of letting deps stay and then cleaning up periodically (e.g. once or twoce a week) is the simpler, more robust approach.

Still, would be nice to have it "just the way I want", removing make deps (only once, at the right time, i.e. after install of final package in the dep tree) without hiccups.

**Comment by Kr1ss-XD** [comment]
Yep this would only work as long as there are no dependencies to be installed from AUR. That's what I was missing, and what @Morganamilo pointed out.

**Comment by pauldotknopf** [comment]
Why not build the package in an Arch-based docker image? Then, put your binaries on GitHub, and add your GitHub repo as a source for pacman?

https://github.com/pauldotknopf/binary-repos

The pacman repo is here: https://github.com/pauldotknopf/binary-repos/releases/tag/arch-linux-release

```
echo -e "[pauldotknopf]\nServer = https://github.com/pauldotknopf/binary-repos/releases/download/arch-linux-release/\nSigLevel = Optional TrustAll" >> /etc/pacman.conf
```

**Comment by Kr1ss-XD** [comment]
Nice approach. I actually do something similar, but I build locally in a chroot, and my repo also is local.

Using a remote repo, you could even easily deploy your packages to several machines.

**Comment by pauldotknopf** [comment]
> Using a remote repo, you could even easily deploy your packages to several machines.

If you want to take it a step further, check out [Darch](https://godarch.com) (I wrote it).

I deploy my operating systems to Docker Hub [here](https://cloud.docker.com/repository/docker/pauldotknopf/darch-ubuntu-development/tags), built from my recipes [here](https://github.com/pauldotknopf/darch-recipes), published by Travis.

All my machines run the same exact OS, bit for bit. When I need to update any one of them, I simply run ```update-dev-image && sudo reboot now``` ([source](https://github.com/pauldotknopf/dotfiles/blob/523635d8dc2ff75a3dce43111544effad7f5852c/zshrc#L47)).

Darch is available [in the AUR](https://aur.archlinux.org/packages/darch/).

Darch also makes it easy to distro-hop. My recipes used to be Arch based, but now they are Ubuntu based. As soon as I ran ```update-dev-image``` after updating my recipes, I was in Ubuntu.

**Comment by stef204** [comment]
> **Edit** Maybe you could install a cron job or a systemd timer/service to clean up every once in a while ?

Could do.  But I  have a shell alias for `trizen -Rcns $(pacman -Qqdt)` or `pacman -Rcns $(pacman -Qqdt)` and I can just run that once in a while; simple enough.  I don't want to automate that type of stuff.
